test/widgets/compose_box_test.dart

```diff
@@ -668,4 +681,81 @@ void main() {
       });
     });
   });
+
+  group('ComposeBox content input scaling', () {
+    const verticalPadding = 8;
+    final stream = eg.stream();
+    final narrow = TopicNarrow(stream.streamId, 'foo');
+
+    Future<void> checkContentInputMaxHeight(WidgetTester tester, {
+      required double maxHeight,
+      required int maxVisibleLines,
+    }) async {
+      final finder = find.byType(ComposeBox);
+
+      double? height;
+      // The content input contains two lines at minimum.
+      String content = 'foo\nfoo';
+
+      // On top of the existing two lines, add exactly the number of lines
+      // needed such that the last line will not visible in the content input
+      // after reaching the max height.
+      for (int i = 0; i < maxVisibleLines - 2 + 1; i++) {
```

Comment:
This loop is kind of confusing to follow. A very local fix would be to make `i` represent the number of lines in `content`:
```suggestion
      // On top of the existing two lines, add exactly the number of lines
      // needed such that the last line will not visible in the content input
      // after reaching the max height.
      for (int i = 2; i < maxVisibleLines + 1; i++) {
```

I think a further clarification is then to have just one loop variable, the number of lines, instead of effectively two in parallel (`i` and `content`). That can look like this:
```dart
      // Add one line at a time, until the content input reaches its max height.
      int numLines;
      double? height;
      for (numLines = 2; numLines <= 1000; numLines++) {
        final content = List.generate(numLines, (_) => 'foo').join('\n');
        await tester.enterText(finder, content);
        await tester.pump();
        final newHeight = tester.getRect(contentInputFinder).height;
        if (newHeight == height) {
          break;
        }
        height = newHeight;
      }
      check(height).isNotNull().isCloseTo(maxHeight, 0.5);
      // The last line added did not stretch the content input,
      // so only the lines before it are at least partially visible.
      check(numLines - 1).equals(maxVisibleLines);
```